Feature: Initial implementation of ButtonDropdown View.

jodator opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Initial implementation of ButtonGroupView and ButtonDropdown View. Closes #333.

---

### Additional information

* This PR mostly required by [alignment feature](https://github.com/ckeditor/ckeditor5-alignment/pull/7) but proposed button dropdown is a generic component.
* I've decided to use Buttons directly to enable [easy button wrapping to button dropdown](https://github.com/ckeditor/ckeditor5-ui/issues/322).

* This PR also requires change in theme lark: https://github.com/ckeditor/ckeditor5-theme-lark/tree/t/ckeditor5-ui/333 with proposed merge commit msg:
```
Other: Make ButtonDropdown's toolbar button present as normal button.
```

oleq commented on src/buttongroup/buttongroupview.js:
```diff
@@ -0,0 +1,128 @@
+ *
+ * @extends module:ui/view~View
+ */
+export default class ButtonGroupView extends View {
```
I wonder if this class could be replaced by the `ToolbarView`? It almost replicates its functionality 1:1. The only feature that `ToolbarView` would need is `isVertical`, right?

jodator replied:
Looks like it is a 1:1 copy with additional `isVertical` property behavior. It would only need some minor CSS adjustments for paddings:

![selection_047](https://user-images.githubusercontent.com/247363/33880057-a3d948c8-df30-11e7-8f7a-b783a15b09b3.png)

The other things is (probably minor) the name of this thing as I don't feel too comfortable with placing Toolbar inside dropdown. But as it works and does not create bloat code I can live with this :)

oleq commented on src/dropdown/button/createbuttondropdown.js:
```diff
@@ -0,0 +1,106 @@
+ *
+ *		const buttons = [];
+ *
+ * 		buttons.push( new ButtonView() );
```
Indent.

oleq commented on src/dropdown/button/createbuttondropdown.js:
```diff
@@ -0,0 +1,106 @@
+	return dropdownView;
+}
+
+function getBindingTargets( buttons, attribute ) {
```
Missing docs.

oleq commented on src/dropdown/utils.js:
```diff
@@ -0,0 +1,66 @@
+
+/* global document */
+
+export function openDropdownOnArrows( dropdownView, buttonGroupView ) {
```
Missing docs.

oleq commented on src/dropdown/utils.js:
```diff
@@ -0,0 +1,66 @@
+	} );
+}
+
+export function closeDropdownOnExecute( dropdownView, items ) {
```
Missing docs.

oleq commented on src/dropdown/utils.js:
```diff
@@ -0,0 +1,66 @@
+	} );
+}
+
+export function closeDropdownOnBlur( dropdownView ) {
```
Missing docs.

oleq commented on tests/dropdown/manual/dropdown.js:
```diff
@@ -95,7 +103,36 @@ function testLongLabel() {
 testList();
 testSharedModel();
 testLongLabel();
+testButton();
```
I'm missing some manual tests here:
* Horizontal buttons
* Dynamic button icon

jodator replied:
I've created this: https://github.com/ckeditor/ckeditor5-ui/blob/e44374384a421862a64e55a05b88594c0ae414c6/tests/dropdown/manual/dropdown.md#button-dropdown so you can fiddle with in in console.
Please let me know if this is enough or should I add some scripts/buttons for that?

oleq commented on theme/components/buttongroup.css:
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
```
This file, if `ButtonGroupView` becomes a component (my comment in the class definition), should have its own folder.

oleq commented on src/dropdown/button/createbuttondropdown.js:
```diff
@@ -0,0 +1,106 @@
+
+	dropdownView.buttonView.extendTemplate( {
+		attributes: {
+			class: [ 'ck-button-dropdown' ]
```
I think classes are set on wrong DOM elements. ATM it's

```html
<div class="ck-dropdown">
    <button class="ck-button ck-enabled ck-off ck-dropdown__button ck-button-dropdown" type="button" tabindex="-1">...</button>
    <div class="ck-reset ck-dropdown__panel">
        <div class="ck-reset ck-button-group ">...</div>
    </div>
</div>
```

but I think it should be 

```html
<div class="ck-dropdown ck-button-dropdown">
    <button class="ck-button ck-enabled ck-off ck-dropdown__button" type="button" tabindex="-1">...</button>
    <div class="ck-reset ck-dropdown__panel">
        <div class="ck-reset ck-button-group ">...</div>
    </div>
</div>
```

instead. WDYT?

oleq commented on src/buttongroup/buttongroupview.js:
```diff
@@ -0,0 +1,128 @@
+	/**
+	 * @inheritDoc
+	 */
+	constructor( options = {} ) {
```
Docs for `options`? `View` does not accept such argument, so `@inheritDoc` won't work.

oleq commented on src/buttongroup/buttongroupview.js:
```diff
@@ -0,0 +1,128 @@
+	 * @inheritDoc
+	 */
+	constructor( options = {} ) {
+		super();
```
I think it's safe to pass `locale` to the parent here. If it's needed in the future, it will require changes in many places that use this component because the order of `constructor()` arguments will change (`locale` should be first).

oleq commented on src/buttongroup/buttongroupview.js:
```diff
@@ -0,0 +1,128 @@
+import '../../theme/components/buttongroup.css';
+
+/**
+ * The button group view class.
```
Missing some simple example.

jodator commented:
@oleq I've updated this PR with:
- I've used `ToolbarView` directly with only one addition of `focusLast()` method.
- I've updated some docs.
- I've created a docs for `DropdownPanelFocusable` as I saw that one method from `dropdown/utils` required a view with `focus()` and `focusLast()` methods.
- Manual tests for button dropdown has now it's own section.
